Algorithms/Knapsack/BranchAndBoundKnapsackSolver.cs

```diff
@@ -0,0 +1,187 @@
+            }
+
+            return GetOptimalItems(items, lastNodeOfOptimalPath);
+        }
+
+        // determine items taken based on the path that gives maximum value
+        private static T[] GetOptimalItems(T[] items, BranchAndBoundNode lastNodeOfOptimalPath)
+        {
+            List<T> takenItems = new List<T>();
+
+            BranchAndBoundNode? currentNode = lastNodeOfOptimalPath;
+            int numberOfNodes = lastNodeOfOptimalPath.GetLevel();
+
+            for (int i = numberOfNodes; i >= 0; i--)
+            {
+                // the node's level's item is taken, add to knapsack taken items list
+                if(currentNode != null)
+                {
+                    if(currentNode.IsTaken())
+                    {
+                        takenItems.Add(items[i]);
+                    }
+
+                    // set currentNode to its parent to check if the parent is taken in the next iteration
+                    currentNode = currentNode.GetParent();
+                }
+            }
+
+            return takenItems.ToArray();
+        }
+
+        /// <summary>
+        ///     Returns the upper bound value of a given node.
+        /// </summary>
+        /// <param name="aNode">The given node.</param>
+        /// <param name="items">All items to choose from.</param>
+        /// <param name="capacity">The maximum weight capacity of the knapsack to be filled.</param>
+        /// <param name="weightSelector">
+        ///     A function that returns the value of the specified item
+        ///     from the <paramref name="items">items</paramref> list.
+        /// </param>
+        /// <param name="valueSelector">
+        ///     A function that returns the weight of the specified item
+        ///     from the <paramref name="items">items</paramref> list.
+        /// </param>
+        /// <returns>
+        ///     upper bound value of the given <paramref name="aNode">node</paramref>.
+        /// </returns>
+        private static double ComputeUpperBound(BranchAndBoundNode aNode, T[] items, int capacity, Func<T, int> weightSelector, Func<T, double> valueSelector)
+        {
+            double upperBound = aNode.GetCumulativeValue();
+            int availableWeight = capacity - aNode.GetCumulativeWeight();
+            int nextLevel = aNode.GetLevel() + 1;
+
+            while (availableWeight > 0 && nextLevel < items.Length)
+            {
+                if (weightSelector(items[nextLevel]) <= availableWeight)
+                {
+                    upperBound += valueSelector(items[nextLevel]);
+                    availableWeight -= weightSelector(items[nextLevel]);
+                }
+                else
+                {
+                    upperBound += valueSelector(items[nextLevel]) / weightSelector(items[nextLevel]) * availableWeight;
+                    availableWeight = 0;
+                }
+
+                nextLevel++;
+            }
+
+            return upperBound;
+        }
```

Comment:
Could you share sources for that? Looks like a heuristic to me.

Reply:
Sorry my English is bad. From what I get from wikipedia "A heuristic algorithm is one that is designed to solve a problem in a faster and more efficient fashion than traditional methods by sacrificing optimality, accuracy, precision, or completeness for speed". This is in fact a more efficient solution when compared to brute force. I watched Bari's explanation regarding this algorithm on <https://www.youtube.com/watch?v=yV1d-b_NeK8>. The upper bound calculation was following what he explained in the video.

Reply:
The key point is 'by sacrificing optimality'. I've found what is missing in the current implementation to give optimal solutions. This is step 1 from https://www.geeksforgeeks.org/implementation-of-0-1-knapsack-using-branch-and-bound/. I've implemented a fix and waiting on you to give me push access to your repo to push updates. (I don't want to make a fork of a fork :) to then make a PR for another PR)